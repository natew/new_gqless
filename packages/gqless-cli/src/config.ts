import { cosmiconfig } from 'cosmiconfig';
import { promises } from 'fs';
import { resolve } from 'path';
import { Boolean, Partial, Static, String, Unknown } from 'runtypes';

import type { GenerateOptions } from './generate';
import type { IntrospectionOptions } from './introspection';

type GqlessCombinedOptions = GenerateOptions & {
  /**
   * Introspection options
   */
  introspection?: IntrospectionOptions;
  /**
   * Client generation destination
   */
  destination?: string;
};

export type GqlessConfig = Required<
  Static<typeof gqlessCLIConfigRecord>
> extends Required<GqlessCombinedOptions>
  ? GqlessCombinedOptions
  : never;

const StringRecord = Unknown.withConstraint<Record<string, string>>(
  (scalars) => {
    return (
      typeof scalars === 'object' &&
      scalars != null &&
      !Array.isArray(scalars) &&
      Object.values(scalars).every((v) => typeof v === 'string')
    );
  }
);

const gqlessCLIConfigRecord = Partial({
  endpoint: String,
  enumsAsStrings: Boolean,
  scalars: StringRecord,
  react: Boolean,
  preImport: String,
  introspection: Partial({
    endpoint: String,
    headers: StringRecord,
  }),
  destination: String,
  subscriptions: Boolean,
});

export const defaultConfig = {
  endpoint: '/api/graphql',
  enumsAsStrings: false,
  react: false,
  scalars: {
    DateTime: 'string',
  },
  preImport: '',
  introspection: {
    endpoint: 'SPECIFY_ENDPOINT_OR_SCHEMA_FILE_PATH_HERE',
    headers: {},
  } as IntrospectionOptions,
  destination: './src/generated/graphql.ts',
  subscriptions: false,
};

const defaultFilePath = resolve(process.cwd(), 'gqless.config.cjs');

type DeepReadonly<T> = T extends (infer R)[]
  ? DeepReadonlyArray<R>
  : T extends Function
  ? T
  : T extends object
  ? DeepReadonlyObject<T>
  : T;

interface DeepReadonlyArray<T> extends ReadonlyArray<DeepReadonly<T>> {}

type DeepReadonlyObject<T> = {
  readonly [P in keyof T]: DeepReadonly<T[P]>;
};

export const gqlessConfigPromise: Promise<{
  filepath: string;
  config: DeepReadonly<GqlessConfig>;
}> = cosmiconfig('gqless', {
  searchPlaces: ['gqless.config.cjs', 'gqless.config.js', 'package.json'],
})
  .search()
  .then(async (config) => {
    if (!config || config.isEmpty) {
      const filepath = config?.filepath || defaultFilePath;

      const NODE_ENV = process.env['NODE_ENV'];

      if (NODE_ENV !== 'test' && NODE_ENV !== 'production') {
        const { format } = (await import('./prettier')).formatPrettier({
          parser: 'typescript',
        });
        await promises.writeFile(
          defaultFilePath,
          await format(`
          /**
           * @type {import("@dish/gqless-cli").GqlessConfig}
           */
          const config = ${JSON.stringify(defaultConfig)};
          
          module.exports = config;`)
        );
      }
      return {
        filepath,
        config: defaultConfig,
      };
    }
    return {
      config: gqlessCLIConfigRecord.check(config.config),
      filepath: config.filepath,
    };
  })
  .catch(() => ({ config: defaultConfig, filepath: defaultFilePath }));
