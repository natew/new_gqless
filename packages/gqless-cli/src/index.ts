export * from "./generate";
export * from "./introspection";
export * from "./writeGenerate";
