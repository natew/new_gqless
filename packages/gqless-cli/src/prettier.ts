import {
  Options as PrettierOptions,
  resolveConfig,
  format as prettierFormat,
} from 'prettier';

const commonConfig = resolveConfig(process.cwd());

export function formatPrettier(
  defaultOptions: Omit<PrettierOptions, 'parser'> &
    Required<Pick<PrettierOptions, 'parser'>>
) {
  const configPromise = commonConfig.then((config) =>
    Object.assign({}, config, defaultOptions)
  );

  return {
    async format(input: string) {
      return prettierFormat(input, await configPromise);
    },
  };
}
