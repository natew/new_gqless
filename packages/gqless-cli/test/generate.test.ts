import { createTestApp, gql } from 'test-utils';

import { generate } from '../src';

test('basic functionality works', async () => {
  const { server, isReady } = createTestApp({
    schema: gql`
      type Query {
        hello: String!
      }
    `,
    resolvers: {
      Query: {
        hello() {
          return 'hello world';
        },
      },
    },
  });

  await isReady;

  const shouldBeIncluded = '// This should be included';

  const { schemaCode, generatedSchema, scalarsEnumsHash } = await generate(
    server.graphql.schema,
    {
      preImport: `
        ${shouldBeIncluded}
        `,
    }
  );

  expect(schemaCode).toMatchSnapshot('generate_code');

  expect(JSON.stringify(generatedSchema, null, 2)).toMatchSnapshot(
    'generate_generatedSchema'
  );

  expect(JSON.stringify(scalarsEnumsHash, null, 2)).toMatchSnapshot(
    'generate_scalarsEnumHash'
  );

  expect(schemaCode.startsWith(shouldBeIncluded)).toBeTruthy();
});

test('custom scalars works', async () => {
  const { server, isReady } = createTestApp({
    schema: gql`
      scalar Custom
      type Query {
        hello: Custom!
      }
    `,
    resolvers: {
      Query: {
        hello() {
          return 'hello world';
        },
      },
    },
  });

  await isReady;

  const { schemaCode, generatedSchema, scalarsEnumsHash } = await generate(
    server.graphql.schema,
    {
      scalars: {
        Custom: '"hello world"',
      },
    }
  );

  expect(schemaCode).toMatchSnapshot('generate_code');

  expect(JSON.stringify(generatedSchema, null, 2)).toMatchSnapshot(
    'generate_customScalars_generatedSchema'
  );

  expect(JSON.stringify(scalarsEnumsHash, null, 2)).toMatchSnapshot(
    'generate_customScalars_scalarsEnumHash'
  );

  expect(
    schemaCode.includes(
      `
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Custom: 'hello world';
}
`.trim()
    )
  ).toBeTruthy();
});

describe('feature complete app', () => {
  const { server, isReady } = createTestApp({
    schema: gql`
      scalar ExampleScalar

      enum GreetingsEnum {
        Hello
        Hi
        Hey
      }
      input GreetingsInput {
        language: String!
        value: String
        scal: ExampleScalar
      }
      type Query {
        simpleString: String!
        stringWithArgs(hello: String!): String!
        stringNullableWithArgs(hello: String!, helloTwo: String): String
        stringNullableWithArgsArray(hello: [String]!): String
        object: Human
        objectArray: [Human]
        objectWithArgs(who: String!): Human!
        arrayString: [String!]!
        arrayObjectArgs(limit: Int): [Human!]!
        greetings: GreetingsEnum!
        giveGreetingsInput(input: GreetingsInput!): String!
        number: Int!
      }
      type Mutation {
        increment(n: Int!): Int!
      }
      type Human {
        name: String!
        father: Human!
        fieldWithArgs(id: Int!): Int!
      }
      type OtherHuman {
        name: String!
      }
      # TODO union support
      union HumanType = Human | OtherHuman
    `,
    resolvers: {},
  });
  beforeAll(async () => {
    await isReady;
  });
  test('generate works', async () => {
    const { schemaCode, generatedSchema, scalarsEnumsHash } = await generate(
      server.graphql.schema
    );

    expect(schemaCode).toMatchSnapshot('featureComplete_code');
    expect(JSON.stringify(generatedSchema)).toMatchSnapshot(
      'featureComplete_generatedSchema'
    );
    expect(JSON.stringify(scalarsEnumsHash)).toMatchSnapshot(
      'featureComplete_scalarsEnumsHash'
    );
  });
});

test('prettier detects invalid code', async () => {
  const { server, isReady } = createTestApp({
    schema: gql`
      type Query {
        hello: String!
      }
    `,
  });
  await isReady;

  await expect(
    generate(server.graphql.schema, {
      preImport: `
        con a; // invalid code
        `,
    })
  ).rejects.toThrow("';' expected. (3:13)");
});

describe('mutation', () => {
  const { server, isReady } = createTestApp({
    schema: gql`
      type Query {
        hello: String!
      }
      type Mutation {
        helloMutation(hello: String!): String!
      }
    `,
    resolvers: {
      Mutation: {
        helloMutation(_root, { hello }: { hello: string }) {
          return hello;
        },
      },
    },
  });

  beforeAll(async () => {
    await isReady;
  });

  test('generates mutation', async () => {
    const { schemaCode, generatedSchema, scalarsEnumsHash } = await generate(
      server.graphql.schema
    );

    expect(schemaCode).toMatchSnapshot('mutation_code');
    expect(generatedSchema).toMatchSnapshot('mutation_generatedSchema');
    expect(scalarsEnumsHash).toMatchSnapshot('mutation_scalarsEnumHash');
  });
});

describe('subscription', () => {
  const { server, isReady } = createTestApp({
    schema: gql`
      type Query {
        hello: String!
      }
      type Subscription {
        newNotification: String!
      }
    `,
    resolvers: {
      Query: {
        hello() {
          return 'hello world';
        },
      },
    },
  });

  beforeAll(async () => {
    await isReady;
  });

  test('generates subscription', async () => {
    const { schemaCode, generatedSchema, scalarsEnumsHash } = await generate(
      server.graphql.schema
    );

    expect(schemaCode).toMatchSnapshot('subscription_code');
    expect(generatedSchema).toMatchSnapshot('subscription_generatedSchema');
    expect(scalarsEnumsHash).toMatchSnapshot('subscription_scalarsEnumHash');
  });
});
