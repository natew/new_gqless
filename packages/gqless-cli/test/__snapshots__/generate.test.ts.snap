// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic functionality works: generate_client_code 1`] = `
"/**
 * GQLESS: You can safely modify this file and Query Fetcher based on your needs
 */

import { createReactClient } from '@dish/gqless-react';
import { createClient, QueryFetcher } from '@dish/gqless';
import {
  GeneratedSchema,
  generatedSchema,
  scalarsEnumsHash,
  SchemaObjectTypes,
  SchemaObjectTypesNames,
} from './schema.generated';

const queryFetcher: QueryFetcher = async function (query, variables) {
  // Modify \\"/api/graphql\\" if needed
  const response = await fetch('/api/graphql', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
    mode: 'cors',
  });

  if (!response.ok) {
    throw new Error(\`Network error, received status code \${response.status}\`);
  }

  const json = await response.json();

  return json;
};

export const client = createClient<
  GeneratedSchema,
  SchemaObjectTypesNames,
  SchemaObjectTypes
>({ schema: generatedSchema, scalarsEnumsHash, queryFetcher });

export const {
  query,
  mutation,
  mutate,
  subscription,
  resolved,
  refetch,
} = client;

export const {
  graphql,
  useQuery,
  useTransactionQuery,
  useLazyQuery,
  usePolling,
  useMutation,
  useRefetch,
  useMetaState,
} = createReactClient<GeneratedSchema>(client, {
  defaults: {
    // Set this flag as \\"false\\" if your usage doesn't involve React Suspense
    suspense: true,

    // Set this flag based on your needs
    staleWhileRevalidate: false,
  },
});

export * from './schema.generated';
"
`;

exports[`basic functionality works: generate_code 1`] = `
"/**
 * GQLESS AUTO-GENERATED CODE: PLEASE DO NOT MODIFY MANUALLY
 */

// This should be included

import { ScalarsEnumsHash } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  String: true,
  Int: true,
  Boolean: true,
};
export const generatedSchema = {
  query: {
    __typename: { __type: 'String!' },
    hello: { __type: 'String!' },
    deprecatedArg: { __type: 'Int', __args: { arg: 'Int' } },
  },
  mutation: {},
  subscription: {},
} as const;

/**
 * Query
 */
export interface Query {
  __typename: 'Query' | undefined;
  /**
   * Hello field
   */
  hello: ScalarsEnums['String'];
  /**
   * @deprecated No longer supported
   */
  deprecatedArg: (args?: {
    /**
     * @defaultValue \`123\`
     */
    arg?: ScalarsEnums['Int'];
  }) => ScalarsEnums['Int'];
}

export interface Mutation {
  __typename: 'Mutation' | undefined;
}

export interface Subscription {
  __typename: 'Subscription' | undefined;
}

export interface SchemaObjectTypes {
  Query: Query;
  Mutation: Mutation;
  Subscription: Subscription;
}
export type SchemaObjectTypesNames = 'Query' | 'Mutation' | 'Subscription';

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export type MakeNullable<T> = {
  [K in keyof T]: T[K] | undefined;
};

export interface ScalarsEnums extends MakeNullable<Scalars> {}
"
`;

exports[`basic functionality works: generate_generatedSchema 1`] = `
"{
  \\"query\\": {
    \\"__typename\\": {
      \\"__type\\": \\"String!\\"
    },
    \\"hello\\": {
      \\"__type\\": \\"String!\\"
    },
    \\"deprecatedArg\\": {
      \\"__type\\": \\"Int\\",
      \\"__args\\": {
        \\"arg\\": \\"Int\\"
      }
    }
  },
  \\"mutation\\": {},
  \\"subscription\\": {}
}"
`;

exports[`basic functionality works: generate_scalarsEnumHash 1`] = `
"{
  \\"String\\": true,
  \\"Int\\": true,
  \\"Boolean\\": true
}"
`;

exports[`custom scalars works: generate_client_code 1`] = `
"/**
 * GQLESS: You can safely modify this file and Query Fetcher based on your needs
 */

import { createClient, QueryFetcher } from '@dish/gqless';
import {
  GeneratedSchema,
  generatedSchema,
  scalarsEnumsHash,
  SchemaObjectTypes,
  SchemaObjectTypesNames,
} from './schema.generated';

const queryFetcher: QueryFetcher = async function (query, variables) {
  // Modify \\"/api/graphql\\" if needed
  const response = await fetch('/api/graphql', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      query,
      variables,
    }),
    mode: 'cors',
  });

  if (!response.ok) {
    throw new Error(\`Network error, received status code \${response.status}\`);
  }

  const json = await response.json();

  return json;
};

export const client = createClient<
  GeneratedSchema,
  SchemaObjectTypesNames,
  SchemaObjectTypes
>({ schema: generatedSchema, scalarsEnumsHash, queryFetcher });

export const {
  query,
  mutation,
  mutate,
  subscription,
  resolved,
  refetch,
} = client;

export * from './schema.generated';
"
`;

exports[`custom scalars works: generate_code 1`] = `
"/**
 * GQLESS AUTO-GENERATED CODE: PLEASE DO NOT MODIFY MANUALLY
 */

import { ScalarsEnumsHash } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  Custom: 'hello world';
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  Custom: true,
  Boolean: true,
  String: true,
};
export const generatedSchema = {
  query: { __typename: { __type: 'String!' }, hello: { __type: 'Custom!' } },
  mutation: {},
  subscription: {},
} as const;

export interface Query {
  __typename: 'Query' | undefined;
  hello: ScalarsEnums['Custom'];
}

export interface Mutation {
  __typename: 'Mutation' | undefined;
}

export interface Subscription {
  __typename: 'Subscription' | undefined;
}

export interface SchemaObjectTypes {
  Query: Query;
  Mutation: Mutation;
  Subscription: Subscription;
}
export type SchemaObjectTypesNames = 'Query' | 'Mutation' | 'Subscription';

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export type MakeNullable<T> = {
  [K in keyof T]: T[K] | undefined;
};

export interface ScalarsEnums extends MakeNullable<Scalars> {}
"
`;

exports[`custom scalars works: generate_customScalars_generatedSchema 1`] = `
"{
  \\"query\\": {
    \\"__typename\\": {
      \\"__type\\": \\"String!\\"
    },
    \\"hello\\": {
      \\"__type\\": \\"Custom!\\"
    }
  },
  \\"mutation\\": {},
  \\"subscription\\": {}
}"
`;

exports[`custom scalars works: generate_customScalars_scalarsEnumHash 1`] = `
"{
  \\"Custom\\": true,
  \\"Boolean\\": true,
  \\"String\\": true
}"
`;

exports[`feature complete app generate works: featureComplete_code 1`] = `
"/**
 * GQLESS AUTO-GENERATED CODE: PLEASE DO NOT MODIFY MANUALLY
 */

import { ScalarsEnumsHash, SchemaUnionsKey } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  ExampleScalar: any;
}

/** Greetings Enum */
export enum GreetingsEnum {
  /** Hello */
  Hello = 'Hello',
  /** Hi */
  Hi = 'Hi',
  /** Hey */
  Hey = 'Hey',
  Bye = 'Bye',
}

export enum OtherEnum {
  Other = 'Other',
}

/** Greetings Input */
export interface GreetingsInput {
  /** Language */
  language: Scalars['String'];
  value?: Maybe<Scalars['String']>;
  scal?: Maybe<Scalars['ExampleScalar']>;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  ExampleScalar: true,
  GreetingsEnum: true,
  OtherEnum: true,
  String: true,
  Int: true,
  Boolean: true,
};
export const generatedSchema = {
  query: {
    __typename: { __type: 'String!' },
    simpleString: { __type: 'String!' },
    stringWithArgs: { __type: 'String!', __args: { hello: 'String!' } },
    stringNullableWithArgs: {
      __type: 'String',
      __args: { hello: 'String!', helloTwo: 'String' },
    },
    stringNullableWithArgsArray: {
      __type: 'String',
      __args: { hello: '[String]!' },
    },
    object: { __type: 'Human' },
    objectArray: { __type: '[Human]' },
    objectWithArgs: { __type: 'Human!', __args: { who: 'String!' } },
    arrayString: { __type: '[String!]!' },
    arrayObjectArgs: { __type: '[Human!]!', __args: { limit: 'Int' } },
    greetings: { __type: 'GreetingsEnum!' },
    giveGreetingsInput: {
      __type: 'String!',
      __args: { input: 'GreetingsInput!' },
    },
    number: { __type: 'Int!' },
  },
  mutation: {
    __typename: { __type: 'String!' },
    increment: { __type: 'Int!', __args: { n: 'Int!' } },
  },
  subscription: {},
  GreetingsInput: {
    language: { __type: 'String!' },
    value: { __type: 'String' },
    scal: { __type: 'ExampleScalar' },
  },
  Human: {
    __typename: { __type: 'String!' },
    name: { __type: 'String!' },
    other: { __type: 'String' },
    father: { __type: 'Human!' },
    fieldWithArgs: { __type: 'Int!', __args: { id: 'Int!' } },
    withArgs: { __type: 'Int', __args: { a: 'Int!', b: 'Int' } },
    withArgs2: { __type: 'Int!', __args: { a: 'Int' } },
  },
  OtherHuman: {
    __typename: { __type: 'String!' },
    name: { __type: 'String!' },
    other: { __type: 'String' },
    withArgs: { __type: 'Int', __args: { a: 'Int!', b: 'Int' } },
    withArgs2: { __type: 'Int!', __args: { a: 'Int' } },
  },
  [SchemaUnionsKey]: { HumanType: ['Human', 'OtherHuman'] },
} as const;

export interface Query {
  __typename: 'Query' | undefined;
  simpleString: ScalarsEnums['String'];
  stringWithArgs: (args: {
    hello: ScalarsEnums['String'];
  }) => ScalarsEnums['String'];
  stringNullableWithArgs: (args: {
    hello: ScalarsEnums['String']
    /**
     * @defaultValue \`\\"Hi\\"\`
     */;
    helloTwo?: ScalarsEnums['String'];
  }) => ScalarsEnums['String'];
  stringNullableWithArgsArray: (args: {
    hello: Array<Maybe<ScalarsEnums['String']>>;
  }) => ScalarsEnums['String'];
  object?: Maybe<Human>;
  objectArray?: Maybe<Array<Maybe<Human>>>;
  objectWithArgs: (args: {
    /**
     * Who?
     */
    who: ScalarsEnums['String'];
  }) => Human;
  arrayString: Array<ScalarsEnums['String']>;
  arrayObjectArgs: (args?: {
    /**
     * @defaultValue \`10\`
     */
    limit?: ScalarsEnums['Int'];
  }) => Array<Human>;
  greetings: ScalarsEnums['GreetingsEnum'];
  giveGreetingsInput: (args: {
    input: GreetingsInput;
  }) => ScalarsEnums['String'];
  number: ScalarsEnums['Int'];
}

export interface Mutation {
  __typename: 'Mutation' | undefined;
  increment: (args: { n: ScalarsEnums['Int'] }) => ScalarsEnums['Int'];
}

export interface Subscription {
  __typename: 'Subscription' | undefined;
}

export interface Human extends NamedEntity {
  __typename: 'Human' | undefined;
  name: ScalarsEnums['String'];
  other?: ScalarsEnums['String'];
  father: Human;
  fieldWithArgs: (args: { id: ScalarsEnums['Int'] }) => ScalarsEnums['Int'];
  withArgs: (args: {
    a: ScalarsEnums['Int'];
    b?: ScalarsEnums['Int'];
  }) => ScalarsEnums['Int'];
  withArgs2: (args?: { a?: ScalarsEnums['Int'] }) => ScalarsEnums['Int'];
}

export interface OtherHuman {
  __typename: 'OtherHuman' | undefined;
  name: ScalarsEnums['String'];
  other?: ScalarsEnums['String'];
  withArgs: (args: {
    a: ScalarsEnums['Int'];
    b?: ScalarsEnums['Int'];
  }) => ScalarsEnums['Int'];
  withArgs2: (args?: { a?: ScalarsEnums['Int'] }) => ScalarsEnums['Int'];
}

export interface SchemaObjectTypes {
  Query: Query;
  Mutation: Mutation;
  Subscription: Subscription;
  Human: Human;
  OtherHuman: OtherHuman;
}
export type SchemaObjectTypesNames =
  | 'Query'
  | 'Mutation'
  | 'Subscription'
  | 'Human'
  | 'OtherHuman';

export type HumanType =
  | {
      __typename: 'Human' | undefined;
      father: Human;
      fieldWithArgs: (args: { id: ScalarsEnums['Int'] }) => ScalarsEnums['Int'];
      name: ScalarsEnums['String'];
      other?: ScalarsEnums['String'];
      withArgs: (args: {
        a: ScalarsEnums['Int'];
        b?: ScalarsEnums['Int'];
      }) => ScalarsEnums['Int'];
      withArgs2: (args?: { a?: ScalarsEnums['Int'] }) => ScalarsEnums['Int'];
    }
  | {
      __typename: 'OtherHuman' | undefined;
      father?: undefined;
      fieldWithArgs?: undefined;
      name: ScalarsEnums['String'];
      other?: ScalarsEnums['String'];
      withArgs: (args: {
        a: ScalarsEnums['Int'];
        b?: ScalarsEnums['Int'];
      }) => ScalarsEnums['Int'];
      withArgs2: (args?: { a?: ScalarsEnums['Int'] }) => ScalarsEnums['Int'];
    };

/**
 * Named Entity
 */
export interface NamedEntity {
  /**
   * Named Entity Name
   */
  name: ScalarsEnums['String'];
  other?: ScalarsEnums['String'];
}

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export type MakeNullable<T> = {
  [K in keyof T]: T[K] | undefined;
};

export interface ScalarsEnums extends MakeNullable<Scalars> {
  GreetingsEnum: GreetingsEnum | undefined;
  OtherEnum: OtherEnum | undefined;
}
"
`;

exports[`feature complete app generate works: featureComplete_generatedSchema 1`] = `"{\\"query\\":{\\"__typename\\":{\\"__type\\":\\"String!\\"},\\"simpleString\\":{\\"__type\\":\\"String!\\"},\\"stringWithArgs\\":{\\"__type\\":\\"String!\\",\\"__args\\":{\\"hello\\":\\"String!\\"}},\\"stringNullableWithArgs\\":{\\"__type\\":\\"String\\",\\"__args\\":{\\"hello\\":\\"String!\\",\\"helloTwo\\":\\"String\\"}},\\"stringNullableWithArgsArray\\":{\\"__type\\":\\"String\\",\\"__args\\":{\\"hello\\":\\"[String]!\\"}},\\"object\\":{\\"__type\\":\\"Human\\"},\\"objectArray\\":{\\"__type\\":\\"[Human]\\"},\\"objectWithArgs\\":{\\"__type\\":\\"Human!\\",\\"__args\\":{\\"who\\":\\"String!\\"}},\\"arrayString\\":{\\"__type\\":\\"[String!]!\\"},\\"arrayObjectArgs\\":{\\"__type\\":\\"[Human!]!\\",\\"__args\\":{\\"limit\\":\\"Int\\"}},\\"greetings\\":{\\"__type\\":\\"GreetingsEnum!\\"},\\"giveGreetingsInput\\":{\\"__type\\":\\"String!\\",\\"__args\\":{\\"input\\":\\"GreetingsInput!\\"}},\\"number\\":{\\"__type\\":\\"Int!\\"}},\\"mutation\\":{\\"__typename\\":{\\"__type\\":\\"String!\\"},\\"increment\\":{\\"__type\\":\\"Int!\\",\\"__args\\":{\\"n\\":\\"Int!\\"}}},\\"subscription\\":{},\\"GreetingsInput\\":{\\"language\\":{\\"__type\\":\\"String!\\"},\\"value\\":{\\"__type\\":\\"String\\"},\\"scal\\":{\\"__type\\":\\"ExampleScalar\\"}},\\"Human\\":{\\"__typename\\":{\\"__type\\":\\"String!\\"},\\"name\\":{\\"__type\\":\\"String!\\"},\\"other\\":{\\"__type\\":\\"String\\"},\\"father\\":{\\"__type\\":\\"Human!\\"},\\"fieldWithArgs\\":{\\"__type\\":\\"Int!\\",\\"__args\\":{\\"id\\":\\"Int!\\"}},\\"withArgs\\":{\\"__type\\":\\"Int\\",\\"__args\\":{\\"a\\":\\"Int!\\",\\"b\\":\\"Int\\"}},\\"withArgs2\\":{\\"__type\\":\\"Int!\\",\\"__args\\":{\\"a\\":\\"Int\\"}}},\\"OtherHuman\\":{\\"__typename\\":{\\"__type\\":\\"String!\\"},\\"name\\":{\\"__type\\":\\"String!\\"},\\"other\\":{\\"__type\\":\\"String\\"},\\"withArgs\\":{\\"__type\\":\\"Int\\",\\"__args\\":{\\"a\\":\\"Int!\\",\\"b\\":\\"Int\\"}},\\"withArgs2\\":{\\"__type\\":\\"Int!\\",\\"__args\\":{\\"a\\":\\"Int\\"}}}}"`;

exports[`feature complete app generate works: featureComplete_scalarsEnumsHash 1`] = `"{\\"ExampleScalar\\":true,\\"GreetingsEnum\\":true,\\"OtherEnum\\":true,\\"String\\":true,\\"Int\\":true,\\"Boolean\\":true}"`;

exports[`mutation generates mutation: mutation_code 1`] = `
"/**
 * GQLESS AUTO-GENERATED CODE: PLEASE DO NOT MODIFY MANUALLY
 */

import { ScalarsEnumsHash } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  String: true,
  Boolean: true,
};
export const generatedSchema = {
  query: { __typename: { __type: 'String!' }, hello: { __type: 'String!' } },
  mutation: {
    __typename: { __type: 'String!' },
    helloMutation: { __type: 'String!', __args: { hello: 'String!' } },
  },
  subscription: {},
} as const;

export interface Query {
  __typename: 'Query' | undefined;
  hello: ScalarsEnums['String'];
}

export interface Mutation {
  __typename: 'Mutation' | undefined;
  helloMutation: (args: {
    hello: ScalarsEnums['String'];
  }) => ScalarsEnums['String'];
}

export interface Subscription {
  __typename: 'Subscription' | undefined;
}

export interface SchemaObjectTypes {
  Query: Query;
  Mutation: Mutation;
  Subscription: Subscription;
}
export type SchemaObjectTypesNames = 'Query' | 'Mutation' | 'Subscription';

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export type MakeNullable<T> = {
  [K in keyof T]: T[K] | undefined;
};

export interface ScalarsEnums extends MakeNullable<Scalars> {}
"
`;

exports[`mutation generates mutation: mutation_generatedSchema 1`] = `
Object {
  "mutation": Object {
    "__typename": Object {
      "__type": "String!",
    },
    "helloMutation": Object {
      "__args": Object {
        "hello": "String!",
      },
      "__type": "String!",
    },
  },
  "query": Object {
    "__typename": Object {
      "__type": "String!",
    },
    "hello": Object {
      "__type": "String!",
    },
  },
  "subscription": Object {},
}
`;

exports[`mutation generates mutation: mutation_scalarsEnumHash 1`] = `
Object {
  "Boolean": true,
  "String": true,
}
`;

exports[`subscription generates subscription: subscription_code 1`] = `
"/**
 * GQLESS AUTO-GENERATED CODE: PLEASE DO NOT MODIFY MANUALLY
 */

import { ScalarsEnumsHash } from '@dish/gqless';

export type Maybe<T> = T | null;
export type Exact<T extends { [key: string]: unknown }> = {
  [K in keyof T]: T[K];
};
export type MakeOptional<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> &
  { [SubKey in K]: Maybe<T[SubKey]> };
/** All built-in and custom scalars, mapped to their actual values */
export interface Scalars {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
}

export const scalarsEnumsHash: ScalarsEnumsHash = {
  String: true,
  Boolean: true,
};
export const generatedSchema = {
  query: { __typename: { __type: 'String!' }, hello: { __type: 'String!' } },
  mutation: {},
  subscription: {
    __typename: { __type: 'String!' },
    newNotification: { __type: 'String!' },
  },
} as const;

export interface Query {
  __typename: 'Query' | undefined;
  hello: ScalarsEnums['String'];
}

export interface Mutation {
  __typename: 'Mutation' | undefined;
}

export interface Subscription {
  __typename: 'Subscription' | undefined;
  newNotification: ScalarsEnums['String'];
}

export interface SchemaObjectTypes {
  Query: Query;
  Mutation: Mutation;
  Subscription: Subscription;
}
export type SchemaObjectTypesNames = 'Query' | 'Mutation' | 'Subscription';

export interface GeneratedSchema {
  query: Query;
  mutation: Mutation;
  subscription: Subscription;
}

export type MakeNullable<T> = {
  [K in keyof T]: T[K] | undefined;
};

export interface ScalarsEnums extends MakeNullable<Scalars> {}
"
`;

exports[`subscription generates subscription: subscription_generatedSchema 1`] = `
Object {
  "mutation": Object {},
  "query": Object {
    "__typename": Object {
      "__type": "String!",
    },
    "hello": Object {
      "__type": "String!",
    },
  },
  "subscription": Object {
    "__typename": Object {
      "__type": "String!",
    },
    "newNotification": Object {
      "__type": "String!",
    },
  },
}
`;

exports[`subscription generates subscription: subscription_scalarsEnumHash 1`] = `
Object {
  "Boolean": true,
  "String": true,
}
`;
